### ISSUE TITLE
Support multiple @Top statements

### ISSUE THREAD

**Issue by nikku** [lezer-parser/lezer#10]
As sketched [in this forum topic](https://discuss.codemirror.net/t/2249) I have a lezer-based grammar written for a domain specific language (cf. [GitHub](https://github.com/nikku/feelin)). The language has two entry points and a long list of shared grammar rules: 

```
@Top[name=Expressions] {
  Expression+
}

@Top[name=UnaryTests] {
  Expression+,
  "not" "(" Expression+ ")",
  "-"
}

/* ... long list of shared grammar definitions */
```

As you mentioned in the forum post lezer does not support multiple `@Top` entry points yet. 

__Describe the solution you'd like__

* The lezer supports grammars with multiple entry points `@Top`
* When parsing an expression, the user must define which root node she would like to start parsing from: 
    
    ```javascript
    parser.parse(expression, { top: `Expressions` });
    parser.parse(expression, { top: `UnaryTests` });
    ```

**Comment by nikku** [comment]
Could you give me basic pointers where to start implementing this feature? 

I'd like to try contributing it.

**Comment by marijnh** [comment]
I guess the first issue would be to look into how states could be reused within different targets. At a glance, it looks like that'd at least require different states to be tagged as accepting, and states do point at each other so anything that transitively points at those end states would need to be duplicated. On the other hand, due to the way reductions pick their successor state from another state further up the stack, maybe only the states that involve the top level rules need to be duplicated.

I guess as a start you could look what happens when you extend `buildFullAutomaton` to take an array of start terms, instead of just a single one, and loop over those. You could then see how much that blows up the state space.

**Comment by nikku** [comment]
Thanks for the pointers. I'll have a look and see how much looking blows my mind :wink:.

**Comment by marijnh** [comment]
I was briefly thinking about doing some work on this last week, but decided that I don't have time. Did you make any progress? I believe it shouldn't be inherently problematic, and might actually be relatively easy to implement. The finnicky part is probably in adjusting the serialized parser representation.

**Comment by nikku** [comment]
Sorry for not getting back to you earlier.

I gave it a couple of hours of research but ended up failing due to my lack of understanding of the parser. Plus, I ~~wasted~~ spend an hour or so on a proper TypeScript dev setup. 

So no, I did not accomplish anything yet.

**Comment by marijnh** [comment]
Oh, no pressure at all, was just curious. (And no, TypeScript isn't really a joy to set up.)

**Comment by nikku** [comment]
I was able to make some progress on the issue:

https://github.com/lezer-parser/lezer-generator/pull/1 implements parsing multiple top rules and a test case.
https://github.com/lezer-parser/lezer/pull/17 adds a `top` property to `ParseOptions` so that a user may specify which top to start parsing from:

```
parser.parse(input, { top: 'OtherTop' });
```

Unfortunately I'm stuck at this point, as I was not able to understand how the parser (`lezer`) identifies top statements. That part (selecting the right top statement) is missing for now.

**Comment by nikku** [comment]
Could you validate that what I came up with goes into the right direction?

**Comment by nikku** [comment]
Nevermind my [previous comment](https://github.com/lezer-parser/lezer/issues/10#issuecomment-578398733). I re-read your suggestion and understood now that the state machine is built in lezer-generator `#buildFullAutomation`. Will see how far I get, making that multi-start aware.

**Comment by nikku** [comment]
Updated https://github.com/lezer-parser/lezer-generator/pull/1 with an attempt to simply parse multiple top rules via `#buildFullAutomation`. Result is endless loops when parsing the grammar. I assume that the parse table generated is simply wrong. I don't know where that is coming from though and will stop exploring now :wink:.

**Comment by marijnh** [comment]
I've finished this in attached patches. There were a few small issues with your code, but on the whole there really wasn't much left to fix or implement.

(Not released yet—I'll tag an 0.8.0 of all the packages at some point, but doesn't seem pressing yet.)

**Comment by nikku** [comment]
I can confirm that this works as expected, with my grammar.

**Comment by nikku** [comment]
> I'll tag an 0.8.0 of all the packages at some point, but doesn't seem pressing yet.

What additional feature do you wait for before cutting a release?

**Comment by marijnh** [comment]
I've released 0.8.0 just now.
